### PR TITLE
Implement retry logic

### PR DESCRIPTION
**Pull request by ridingwolf**

### CODE REVIEW

File: src/Be.Vlaanderen.Basisregisters.Projector/Internal/RetryPolicies/ExponentialBackOff.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+                    .WaitAndRetryAsync(_numberOfRetries, CalculateWaitTime)
+                    .ExecuteAsync(async ct => await messageHandler.HandleAsync(messages, ct), token);
+            }
+
+            return  CreateMessageHandlerFor(ExecuteAndRetry);
+        }
+
+        private TimeSpan CalculateWaitTime(int attempt)
+        {
+            return TimeSpan.FromMilliseconds(attempt ^ _wait.Milliseconds);
```
Comment (CumpsD) [inline src/Be.Vlaanderen.Basisregisters.Projector/Internal/RetryPolicies/ExponentialBackOff.cs:39]: Any change to explain this a bit in understandable words? :p

**Comment by CumpsD** [comment]
Looks a good start, some thing which will probably become obvious on usage:

Does the move from `ConnectedProjectionMessageHandler<TContext>` to `IConnectedProjectionMessageHandler` have external impact?

Does the move from 1 message to a list of message have an external impact?

File: src/Be.Vlaanderen.Basisregisters.Projector/Internal/RetryPolicies/ExponentialBackOff.cs (hunk 1 above)
Comment (ridingwolf) [reply src/Be.Vlaanderen.Basisregisters.Projector/Internal/RetryPolicies/ExponentialBackOff.cs:39]: Need to become Math.Pow(...) :)

Comment (CumpsD) [reply src/Be.Vlaanderen.Basisregisters.Projector/Internal/RetryPolicies/ExponentialBackOff.cs]: Reminder after chat: Turn ExponentialBackOff.cs into LinearBackoff.cs

File: src/Be.Vlaanderen.Basisregisters.Projector/Internal/RetryPolicies/LinearBackOff.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using ConnectedProjections;
+    using Microsoft.Extensions.Logging;
+    using Polly;
+    using SqlStreamStore.Streams;
+
+    internal class LinearBackOff<TException> : MessageHandlingRetryPolicy where TException : Exception
```
Comment (CumpsD) [inline src/Be.Vlaanderen.Basisregisters.Projector/Internal/RetryPolicies/LinearBackOff.cs:12]: You can use the Wikipedia style of the word "Backoff": https://en.wikipedia.org/wiki/Exponential_backoff

`LinearBackoff`

File: src/Be.Vlaanderen.Basisregisters.Projector/ConnectedProjections/MessageHandlingRetryPolicy.cs [hunk 3]
```diff
@@ -6,23 +6,38 @@ namespace Be.Vlaanderen.Basisregisters.Projector.ConnectedProjections
...
+            {
+                _messageHandling = messageHandling ?? throw new ArgumentNullException(nameof(messageHandling));
+                RunnerName = projectionName ?? throw new ArgumentNullException(nameof(projectionName));
+                Logger = messageHandlerLogger ?? throw new ArgumentNullException(nameof(messageHandlerLogger));
+            }
 
             public async Task HandleAsync(IEnumerable<StreamMessage> messages, CancellationToken cancellationToken)
                 => await _messageHandling(messages, cancellationToken);
+
+            public ConnectedProjectionName RunnerName { get; }
```
Comment (CumpsD) [inline src/Be.Vlaanderen.Basisregisters.Projector/ConnectedProjections/MessageHandlingRetryPolicy.cs:39]: Props above ctor in our naming style

File: src/Be.Vlaanderen.Basisregisters.Projector/Internal/RetryPolicies/LinearBackOff.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+            async Task ExecuteWithRetryPolicy(IEnumerable<StreamMessage> messages, CancellationToken token)
+            {
+                await Policy
+                    .Handle<TException>()
+                    .WaitAndRetryAsync(
+                        _numberOfRetries,
+                        attempt => _initialWait.Multiply(attempt),
+                        LogRetryAttempt
+                    )
```
Comment (CumpsD) [inline src/Be.Vlaanderen.Basisregisters.Projector/Internal/RetryPolicies/LinearBackOff.cs:48]: Can put the brace after `LogRetryAttempt` in this projects naming style.

**Comment by CumpsD** [comment]
Looking good! Don't forget the filename `LinearBackOff.cs` as well -> `LinearBackoff.cs`

File: test/Be.Vlaanderen.Basisregisters.Projector.Tests/RetryPolicy/LinearBackoffPolicyTests.cs [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+    using Microsoft.Extensions.Logging;
+    using Moq;
+    using SqlStreamStore.Streams;
+    using Xunit;
+
+    public class WhenApplyingALinearBackoffPolicyOnAHandler
+    {
+        private readonly MessageHandlerWithExecutionTracking _handlerWithoutPolicy;
+        private readonly IFixture _fixture;
+        private IConnectedProjectionMessageHandler SUT { get; }
```
Comment (CumpsD) [inline test/Be.Vlaanderen.Basisregisters.Projector.Tests/RetryPolicy/LinearBackoffPolicyTests.cs:23]: usually we write `sut` or `_sut` I think

File: test/Be.Vlaanderen.Basisregisters.Projector.Tests/RetryPolicy/LinearBackoffPolicyTests.cs [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+            { }
+
+            _handlerWithoutPolicy.VerifyExecuted(_messages, Times.Once());
+        }
+    }
+
+    public class WhenMessageHandlerThrowsAnExceptionThatWasNotDefinedToRetryAfterRetrying
+    {
+        private readonly MessageHandlerWithExecutionTracking _handlerWithoutPolicy;
+        private readonly IConnectedProjectionMessageHandler _SUT;
```
Comment (CumpsD) [inline test/Be.Vlaanderen.Basisregisters.Projector.Tests/RetryPolicy/LinearBackoffPolicyTests.cs:119]: `_sut` seems to be the winner :)

File: test/Be.Vlaanderen.Basisregisters.Projector.Tests/RetryPolicy/LinearBackoffPolicyTests.cs [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+        public async Task ThenHandleActionCallsTheHandlerWithoutAPolicyForEachAttempt()
+        {
+            try
+            {
+                await Act();
+            }
+            catch (DoNotRetryException)
+            { }
+
+            _handlerWithoutPolicy.VerifyExecuted(_messages, Times.Exactly(_exceptionSequence.Length));
```
Comment (CumpsD) [inline test/Be.Vlaanderen.Basisregisters.Projector.Tests/RetryPolicy/LinearBackoffPolicyTests.cs:169]: replace this Length with something that means more, 2 or expectedRetries.

**Pull request merged**